src/arch/src/aarch64/gic/icc_regs.rs

```diff
@@ -0,0 +1,211 @@
+SYS_ICC_AP0Rn_EL1!(SYS_ICC_AP0R2_EL1, 2);
+SYS_ICC_AP0Rn_EL1!(SYS_ICC_AP0R3_EL1, 3);
+SYS_ICC_AP1Rn_EL1!(SYS_ICC_AP1R0_EL1, 0);
+SYS_ICC_AP1Rn_EL1!(SYS_ICC_AP1R1_EL1, 1);
+SYS_ICC_AP1Rn_EL1!(SYS_ICC_AP1R2_EL1, 2);
+SYS_ICC_AP1Rn_EL1!(SYS_ICC_AP1R3_EL1, 3);
+
+static VGIC_ICC_REGS: &[u64] = &[
+    SYS_ICC_SRE_EL1,
+    SYS_ICC_CTLR_EL1,
+    SYS_ICC_IGRPEN0_EL1,
+    SYS_ICC_IGRPEN1_EL1,
+    SYS_ICC_PMR_EL1,
+    SYS_ICC_BPR0_EL1,
+    SYS_ICC_BPR1_EL1,
+    SYS_ICC_AP0R0_EL1,
+    SYS_ICC_AP0R1_EL1,
+    SYS_ICC_AP0R2_EL1,
+    SYS_ICC_AP0R3_EL1,
+    SYS_ICC_AP1R0_EL1,
+    SYS_ICC_AP1R1_EL1,
+    SYS_ICC_AP1R2_EL1,
+    SYS_ICC_AP1R3_EL1,
+];
+
+// Helps with triggering either a register fetch or a store.
+enum Action<'a> {
+    Set(&'a [u32], usize),
+    Get(&'a mut Vec<u32>),
+}
+
+#[allow(clippy::trivially_copy_pass_by_ref)]
+fn access_icc_attr(fd: &DeviceFd, offset: u64, typer: u64, val: &u32, set: bool) -> Result<()> {
+    let mut gic_icc_attr = kvm_bindings::kvm_device_attr {
+        group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_CPU_SYSREGS,
+        attr: ((typer & KVM_DEV_ARM_VGIC_V3_MPIDR_MASK) | offset), // this needs the mpidr
+        addr: val as *const u32 as u64,
+        flags: 0,
+    };
+    if set {
+        fd.set_device_attr(&gic_icc_attr).map_err(|e| {
+            Error::DeviceAttribute(e, true, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_REDIST_REGS)
+        })?;
+    } else {
+        fd.get_device_attr(&mut gic_icc_attr).map_err(|e| {
+            Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_REDIST_REGS)
+        })?;
+    }
+    Ok(())
+}
+
+/// Get ICC registers.
+fn access_icc_reg_list(fd: &DeviceFd, gicr_typer: &[u64], action: &mut Action) -> Result<()> {
+    // We need this for the ICC_AP<m>R<n>_EL1 registers.
+    let mut num_priority_bits = 0;
+
+    for i in gicr_typer {
+        for icc_offset in VGIC_ICC_REGS {
+            // As per ARMv8 documentation: https://static.docs.arm.com/ihi0069/c/IHI0069C_
+            // gic_architecture_specification.pdf
+            // page 178,
+            // ICC_AP0R1_EL1 is only implemented in implementations that support 6 or more bits of
+            // priority.
+            // ICC_AP0R2_EL1 and ICC_AP0R3_EL1 are only implemented in implementations that support
+            // 7 bits of priority.
+            if (*icc_offset == SYS_ICC_AP0R1_EL1 || *icc_offset == SYS_ICC_AP1R1_EL1)
+                && num_priority_bits < 6
+            {
+                continue;
+            }
+            if (*icc_offset == SYS_ICC_AP0R2_EL1
+                || *icc_offset == SYS_ICC_AP0R3_EL1
+                || *icc_offset == SYS_ICC_AP1R2_EL1
+                || *icc_offset == SYS_ICC_AP1R3_EL1)
+                && num_priority_bits != 7
+            {
+                continue;
+            }
+            let val;
+            match action {
+                Action::Set(state, idx) => {
+                    val = state[*idx];
```

Comment:
Nit: I would avoid the unnecessary allocation. We can just do `access_icc_attr(fd, *icc_offset, *i, &state[*idx], true)?;`.